### PR TITLE
RUBY-2974 Add FLE 2 API to AutoEncryptionOpts

### PR DESCRIPTION
**Pull request by comandeo**

### CODE REVIEW

File: lib/mongo/client.rb [hunk 1]
```diff
@@ -446,10 +446,18 @@ def hash
     #       JSON Schemas obtained from the server. It protects against a
     #       malicious server advertising a false JSON Schema, which could trick
     #       the client into sending unencrypted data that should be encrypted.
+    #     - Note: If a collection is present on both the encryptedFieldsMap
```
Comment (p-mongo) [inline lib/mongo/client.rb:449]: `in both the :encrypted_fields_map and :schema_map`

**Review by Neilshweky — COMMENTED** [review]
The drivers ticket mentions some Test changes: syncing spec tests and adding encrypted_fields... I don't see them in this PR, why?

File: lib/mongo/client.rb [hunk 2]
```diff
@@ -446,10 +446,18 @@ def hash
...
+    #       and schemaMap, an error will be raised.
     #   - :bypass_auto_encryption => Boolean, when true, disables auto encryption;
     #     defaults to false.
     #   - :extra_options => Hash | nil, options related to spawning mongocryptd
     #     (this part of the API is subject to change).
+    #   - :encrypted_fields_map => Hash | nil, maps a collection namespace to
+    #     an encryptedFields.
```
Comment (p-mongo) [inline lib/mongo/client.rb:456]: encryptedFields is plural and cannot go with a/an, in this case no article would be used.

However as a user unfamiliar with the feature I don't understand what this is saying, is this a copy-paste from somewhere? Can this be linked to server/etc. documentation?

**Review by Neilshweky — APPROVED** [review]

**Comment by comandeo** [comment]
> The drivers ticket mentions some Test changes: syncing spec tests and adding encrypted_fields... I don't see them in this PR, why?

In order to keep PRs small and therefore reviewable, I'm splitting the task into a series of PRs.

File: lib/mongo/client.rb (hunk 1 above)
Comment (comandeo) [reply lib/mongo/client.rb]: Fixed

File: lib/mongo/crypt/binding.rb [hunk 3]
```diff
@@ -1192,6 +1192,110 @@ def self.setopt_crypto_hook_sign_rsaes_pkcs1_v1_5(
...
+      # @!method self.mongocrypt_setopt_encrypted_field_config_map(crypt, efc_map)
+      #   @api private
+      #
+      # Set a local EncryptedFieldConfigMap for encryption.
+      #
+      # @param [ FFI::Pointer ] crypt A pointer to a mongocrypt_t object.
+      # @param [ FFI::Pointer ] efc_map A pointer to mongocrypt_binary_t object that
+      # references a BSON document representing the EncryptedFieldConfigMap
+      # supplied by the user. The keys are collection namespaces and values are
+      # EncryptedFieldConfigMap documents. The viewed data copied. It is valid to
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/crypt/binding.rb]: I think this part is actually irrelevant; removed.

File: lib/mongo/crypt/binding.rb [hunk 4]
```diff
@@ -1192,6 +1192,110 @@ def self.setopt_crypto_hook_sign_rsaes_pkcs1_v1_5(
...
+              data_p
+            )
+          end
+        end
+      end
+
+      # @!method self.mongocrypt_setopt_bypass_query_analysis(crypt)
+      #   @api private
+      #
+      # Opt-into skipping query analysis.
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/crypt/binding.rb]: Fixed.

File: lib/mongo/crypt/handle.rb [hunk 5]
```diff
@@ -93,6 +106,26 @@ def set_schema_map
         Binding.setopt_schema_map(self, @schema_map)
       end
 
+      def set_encrypted_fields_map
+        unless @encrypted_fields_map.is_a?(Hash)
+          raise ArgumentError.new(
+            "#{@encrypted_fields_map} is an invalid encrypted_fields_map; must be a Hash or nil"
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/crypt/handle.rb]: Fixed

File: lib/mongo/crypt/binding.rb [hunk 6]
```diff
@@ -1192,6 +1192,110 @@ def self.setopt_crypto_hook_sign_rsaes_pkcs1_v1_5(
...
+      # If opted in:
+      # - The csfle shared library will not attempt to be loaded.
+      # - A mongocrypt_ctx_t will never enter the MONGOCRYPT_CTX_NEED_MARKINGS state.
+      #
+      # @param [ Mongo::Crypt::Handle ] handle
+      def self.setopt_bypass_query_analysis(handle)
+        mongocrypt_setopt_bypass_query_analysis(handle.ref)
+      end
+
+      # @!method self.mongocrypt_setopt_aes_256_ctr(crypt, aes_256_ctr_encrypt, aes_256_ctr_decrypt, ctx=nil)
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/crypt/binding.rb]: Fixed.

**Review by p-mongo — APPROVED** [review]

**Pull request merged**